aws-s3-transfer-manager/src/operation/download/discovery.rs

```diff
@@ -108,67 +97,33 @@ pub(super) async fn discover_obj(
     Ok(discovery)
 }
 
-async fn discover_obj_with_head(
-    ctx: &DownloadContext,
-    input: &DownloadInput,
-    byte_range: Option<ByteRange>,
-) -> Result<ObjectDiscovery, crate::error::Error> {
-    let resp = ctx
-        .client()
-        .head_object()
-        .set_bucket(input.bucket().map(str::to_string))
-        .set_key(input.key().map(str::to_string))
-        .send()
-        .await
-        .map_err(error::discovery_failed)?;
-    let object_meta: ObjectMetadata = resp.into();
-
-    let remaining = match byte_range {
-        Some(range) => match range {
-            ByteRange::Inclusive(start, end) => start..=end,
-            ByteRange::AllFrom(start) => start..=object_meta.content_length(),
-            ByteRange::Last(n) => {
-                (object_meta.content_length() - n + 1)..=object_meta.content_length()
-            }
-        },
-        None => 0..=object_meta.content_length(),
-    };
-
-    Ok(ObjectDiscovery {
-        remaining,
-        chunk_meta: None,
-        object_meta,
-        initial_chunk: None,
-    })
-}
-
-async fn discover_obj_with_get(
-    ctx: &DownloadContext,
-    input: GetObjectInputBuilder,
+// Use 1..=0 to represent an empty object, and clippy doesn't allow that.
+#[allow(clippy::reversed_empty_ranges)]
+fn discover_handler_first_chunk_response(
+    mut resp: GetObjectOutput,
     range: Option<RangeInclusive<u64>>,
 ) -> Result<ObjectDiscovery, error::Error> {
-    let resp = input.send_with(ctx.client()).await;
-
-    if resp.is_err() {
-        // TODO(aws-sdk-rust#1159) - deal with empty file errors, see https://github.com/awslabs/aws-c-s3/blob/v0.5.7/source/s3_auto_ranged_get.c#L147-L153
-    }
-
-    let mut resp = resp.map_err(error::discovery_failed)?;
-
-    // take the body so we can convert the metadata
     let empty_stream = ByteStream::new(SdkBody::empty());
     let body = mem::replace(&mut resp.body, empty_stream);
-
     let object_meta: ObjectMetadata = (&resp).into();
     let chunk_meta: ChunkMetadata = resp.into();
-    let content_len = chunk_meta.content_length.expect("expected content_length") as u64;
+    let chunk_content_len = chunk_meta
+        .content_length
+        .expect("expected content_length in chunk") as u64;
 
     let remaining = match range {
-        Some(range) => (*range.start() + content_len)..=*range.end(),
-        None => content_len..=object_meta.content_length() - 1,
+        Some(range) => (*range.start() + chunk_content_len)..=*range.end(),
+        None => {
+            if object_meta.content_length() > chunk_content_len {
+                chunk_content_len..=object_meta.content_length() - 1
+            } else {
+                // Return empty range.
+                1..=0
```

Comment:
since `remaining` represents the literal range of data to fetch remaining (as opposed to the amount of data left) I think you can just collapse this to:

```rs
None => chunk_content_len..=(object_meta.content_length() - 1)
```

If the chunk content length is the full object length then the resulting range would be empty.

Reply:
Aaron's suggestion sounds good.

I'm really confused how `1..=0` is an "empty range". It seems like a "bad range" to me?